apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Span.java

```diff
@@ -284,14 +284,14 @@ protected void afterEnd() {
                 return;
             }
             if (!buffered.tryToCompress(this)) {
+                parent.decrementReferences();
                 if (parent.bufferedSpan.compareAndSet(buffered, this)) {
                     this.tracer.endSpan(buffered);
                 } else {
                     // the failed update would ideally lead to a compression attempt with the new buffer,
                     // but we're dropping the compression attempt to keep things simple
                     this.tracer.endSpan(this);
```

Comment:
[for reviewer] `endSpan(this)` might indirectly trigger span recycling, which in turns makes `parent` to be set to `null`, hence triggering an NPE.